required/latex-lab/latex-lab-everymath.dtx

```diff
@@ -0,0 +1,347 @@
+%
+% \subsection{Content grabbing}
+%
+% \begin{macro}{\@@_grab_inline_dollar:w}
+%   Grab up to a single |$|, for inline math mode.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_inline_dollar:w % $
+  #1 $
+  {
+    \@@_process:n {#1} % $
+    #1 $
+  }
+%    \end{macrocode}
+% \end{macro}
+% \begin{macro}{\@@_grab_inline_cmd:w}
+%   The same but for the \LaTeX{} structure.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_inline_cmd:w % \(
+  #1 \)
+  {
+    \@@_process:n {#1} % \(
+    #1 \)
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\@@_grab_dollardollar:w}
+%   And for the classical \TeX{} display structure.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_dollardollar:w % $$
+  #1 $$
+  {
+    \@@_process:n {#1} % $$
+    #1 $$
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \subsection{Document commands}
+%
+% \begin{macro}{\(, \)}
+%   Here, we need to add the boolean around the structure; we also need to
+%   ensure that the conditional is finished before the |$| activates. To do
+%   that, we switch to an \pkg{expl3} structure for ease. There's no downside
+%   here to engine-protected, so we also simplify the setup a little.
+%    \begin{macrocode}
+\cs_gset_protected:Npn \( % \)
+  {
+    \bool_set_true:N \l_@@_cmd_bool
+    \mode_if_math:TF
+      { \@badmath }
+      { $ }
+  } % \(
+\cs_gset_protected:Npn \) %
+  {
+    \mode_if_math:TF
+      { $ }
+      { \@badmath }
+    \bool_set_false:N \l_@@_cmd_bool
+  }
+\cs_undefine:c { ( \c_space_tl }
+\cs_undefine:c { ) \c_space_tl }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}
+%   {\equation, \@@_equation_begin:, \equation*, \@@_equation_star_begin:}
+% \begin{macro}
+%   {\endequation, \@@_equation_end:, \endequation*, \@@_equation_star_end:}
+%   These environments are not set up by \pkg{amsmath} to collect their body,
+%   so we do that here. This has to be done \emph{after} we can be sure
+%   \pkg{amsmath} is loaded.
+%    \begin{macrocode}
+\AddToHook { begindocument }
+  {
+    \cs_new_eq:NN \@@_equation_begin: \equation
+    \cs_new_eq:NN \@@_equation_end: \endequation
+    \RenewDocumentEnvironment { equation } { b }
+      {
+        \bool_set_true:N \l_@@_env_bool
+        \@@_process:n {#1}
+        \@@_equation_begin: #1 \@@_equation_end:
+      }
+      { }
+    \cs_new_eq:Nc \@@_equation_star_begin: { equation* }
+    \cs_new_eq:Nc \@@_equation_star_end: { endequation* }
+    \RenewDocumentEnvironment { equation* } { b }
+      {
+    \bool_set_true:N \l_@@_env_bool
+        \@@_process:n {#1}
+        \@@_equation_star_begin: #1 \@@_equation_star_end:
+      }
+      { }
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \begin{macro}{\[, \]}
+%   A slight oddity: we have to collect the body but want to pass on
+%   to the more formal structure. That means that \cs{]} is no longer
+%   a command, but rather a marker. Again, we need to watch for when
+%   \pkg{amsmath} is loaded after this code.
```

Comment:
Again it would be nice if that could handle `\[`and `\]` being hidden inside an environment

Reply:
Does anyone ever do that? With `amsmath` loaded it's equivalent to `\begin{equation*}` in any case: I'd love to see a real use case here.

Reply:
I can't claim I have a good example taken from the wild but I know I have misued math in the past , for example to get some separator, e.g.
```
\newenvironment{separator}{..some setup... \[}{\]}
```
used as
```
\begin{separator}  * \qquad * \qquad * \end{separator}
```
I'm not claiming such misused has to be supported :-)

but you can think of simply having  different "names" for displays for clarity in the source

Reply:
OK, doable if we have `\[` check the 'environment' flag. If it's not set, then we do a collection step. If it's set, then we assume that a wrapper env. has already done the job.

Reply:
possibly. I'm not claiming any of this has to be done, more like we should at least add notes that this or that may need extending in one way or the other. The real test will be to see how much (if anything) is falling over if, say, the arXiv corpus is tried.

Reply:
I've adjusted the code to allow for this: as noted in the commit message, it currently needs the internal bool setting, so we'll need to decide how to handle that.